From db2615f2b5f5b4d28d31d59df933b52c1faa2b02 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tor=20Arne=20Vestb=C3=B8?= <tor.arne.vestbo@qt.io>
Date: Wed, 22 Oct 2025 15:18:20 +0200
Subject: [PATCH 512/553] iOS: Guard QIOSDocumentPickerController outliving
 QIOSFileDialog

The document picker may outlive the QIOSFileDialog even if we've
released it, if the system has retained it, and we might get delayed
callbacks to e.g. documentPicker:didPickDocumentsAtURLs, in which
case we'd crash.

Pick-to: 6.8 6.5
Change-Id: Idc9f2f6d297332b91d25dc768d1defaf5bde8076
Reviewed-by: Doris Verria <doris.verria@qt.io>
(cherry picked from commit c49419d44b9f2fa015bdcc0ad2ab09587a5a609a)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 .../platforms/ios/qiosdocumentpickercontroller.mm   | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/src/plugins/platforms/ios/qiosdocumentpickercontroller.mm b/src/plugins/platforms/ios/qiosdocumentpickercontroller.mm
index 09e2f2f4c3f..36371a2a8af 100644
--- a/src/plugins/platforms/ios/qiosdocumentpickercontroller.mm
+++ b/src/plugins/platforms/ios/qiosdocumentpickercontroller.mm
@@ -6,8 +6,10 @@
 
 #include "qiosdocumentpickercontroller.h"
 
+#include <QtCore/qpointer.h>
+
 @implementation QIOSDocumentPickerController {
-    QIOSFileDialog *m_fileDialog;
+    QPointer<QIOSFileDialog> m_fileDialog;
 }
 
 - (instancetype)initWithQIOSFileDialog:(QIOSFileDialog *)fileDialog
@@ -60,6 +62,9 @@
 {
     Q_UNUSED(controller);
 
+    if (!m_fileDialog)
+        return;
+
     QList<QUrl> files;
     for (NSURL* url in urls)
         files.append(QUrl::fromNSURL(url));
@@ -70,12 +75,18 @@
 
 - (void)documentPickerWasCancelled:(UIDocumentPickerViewController *)controller
 {
+    if (!m_fileDialog)
+        return;
+
     Q_UNUSED(controller);
     emit m_fileDialog->reject();
 }
 
 - (void)presentationControllerDidDismiss:(UIPresentationController *)presentationController
 {
+    if (!m_fileDialog)
+        return;
+
     Q_UNUSED(presentationController);
 
     // "Called on the delegate when the user has taken action to dismiss the
-- 
2.51.2

